src/iceberg/base_transaction.h

```diff
@@ -0,0 +1,134 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "iceberg/table_identifier.h"
+#include "iceberg/table_requirement.h"
+#include "iceberg/table_update.h"
+#include "iceberg/transaction.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+/// \brief Base class for transaction implementations
+class ICEBERG_EXPORT BaseTransaction : public Transaction {
+ public:
+  BaseTransaction(std::shared_ptr<const Table> table, std::shared_ptr<Catalog> catalog);
+  ~BaseTransaction() override = default;
+
+  const std::shared_ptr<const Table>& table() const override;
+
+  std::unique_ptr<::iceberg::UpdateProperties> UpdateProperties() override;
+
+  std::unique_ptr<AppendFiles> NewAppend() override;
+
+  Status CommitTransaction() override;
+
+  /// \brief Stage updates to be applied upon commit
+  ///
+  /// \param identifier the table identifier
+  /// \param requirements the list of table requirements to validate
+  /// \param updates the list of table updates to apply
+  /// \return a new Table instance with staged updates applied
+  Result<std::unique_ptr<Table>> StageUpdates(
+      const TableIdentifier& identifier,
+      std::vector<std::unique_ptr<TableRequirement>> requirements,
+      std::vector<std::unique_ptr<TableUpdate>> updates);
+
+  /// \brief Whether the last operation has been committed
+  ///
+  /// \return true if the last operation was committed, false otherwise
+  bool HasLastOperationCommitted() const { return context_.last_operation_committed; }
+
+  /// \brief Mark the last operation as committed or not
+  ///
+  /// \param committed true if the last operation was committed, false otherwise
+  void SetLastOperationCommitted(bool committed) {
+    context_.last_operation_committed = committed;
+  }
+
+ protected:
+  /// \brief Apply a list of table updates to the current metadata
+  ///
+  /// \param updates the list of table updates to apply
+  /// \return Status::OK if the updates were applied successfully, or an error status
+  Status ApplyUpdates(const std::vector<std::unique_ptr<TableUpdate>>& updates);
+
+  /// \brief Create a new table update if the last operation has been committed
+  ///
+  /// \param args the arguments to forward to the update constructor
+  /// \return a new UpdateType instance or an error status
+  template <typename UpdateType, typename... Args>
+  Result<std::unique_ptr<UpdateType>> CheckAndCreateUpdate(Args&&... args) {
```

Comment:
This is over-complicated. Let's switch to `std::shared_ptr` for these update types.